doc/coq.rst

```diff
@@ -0,0 +1,392 @@
+command has been deprecated in Coq 8.12.
+
+.. _coq-pp:
+
+coq.pp
+------
+
+Authors of Coq plugins often need to write ``.mlg`` files to extend the Coq
+grammar. Such files are preprocessed with the ``coqpp`` binary. To help plugin
+authors avoid writing boilerplate, we provide a ``(coqpp ...)`` stanza:
+
+.. code:: scheme
+
+    (coq.pp (modules <mlg_list>))
+
+which, for each ``g_mod`` in ``<mlg_list>``, is equivalent to the following
+rule:
+
+.. code:: lisp
+
+    (rule
+     (targets g_mod.ml)
+     (deps (:mlg-file g_mod.mlg))
+     (action (run coqpp %{mlg-file})))
+
+.. _examples:
+
+Examples of Coq Projects
+------------------------
+
+Here we list some examples of some basic Coq project setups in order.
+
+.. _example-simple:
+
+Simple Coq Project
+~~~~~~~~~~~~~~~~~~
+
+Let us start with a simple project. First we make sure we have a
+:ref:`dune-project` file with a :ref:`Coq lang<coq-lang>` stanza present:
+
+.. code:: scheme
+
+  (lang dune 3.2)
+  (using coq 0.3)
+
+Next we need a :ref:`dune file<dune-files>` with a :ref:`coq-theory` stanza:
+
+.. code:: scheme
+  
+  (coq.theory
+   (name myTheory))
+
+
+Finally, we need a Coq ``.v`` file which we name ``A.v``:
+
+
+.. code:: coq
+
+  (** This is my def *)
+  Definition mydef := nat.
+
+Now we run ``dune build``. After this is complete, we get the following files:
+
+.. code::
+
+  .
+  ├── A.v
+  ├── _build
+  │   ├── default
+  │   │   ├── A.glob
+  │   │   ├── A.v
+  │   │   ├── A.v.d
+  │   │   └── A.vo
+  │   └── log
+  ├── dune
+  └── dune-project
+
+.. _example-multi-theory:
+
+Multi-Theory Project
+~~~~~~~~~~~~~~~~~~~~
+
+Here is an example of a more complicated setup:
+
+.. code::
+
+  .
+  ├── A
+  │   ├── AA
+  │   │   └── aa.v
+  │   ├── AB
+  │   │   └── ab.v
+  │   └── dune
+  ├── B
+  │   ├── b.v
+  │   └── dune
+  └── dune-project
+
+Here are the :ref:`dune files<dune-files>`:
+
+.. code:: scheme
+
+  ; A/dune
+  (include_subdirs qualified)
+  (coq.theory
+   (name A))
+
+  ; B/dune
+  (coq.theory
+   (name B)
+   (theories A))
+
+Notice the ``theories`` field in ``B`` allows one :ref:`coq-theory` to depend on
+another. Another thing to note is the inclusion of the :ref:`include_subdirs`
+stanza. This allows our theory to have :ref:`multiple
+subdirectories<include-subdirs-coq>`. 
+
+Here are the contents of the ``.v`` files:
+
+.. code:: coq
+  
+  (* A/AA/aa.v is empty *)
+
+  (* A/AB/ab.v *)
+  Require Import AA.aa.
+
+  (* B/b.v *)
+  From A Require Import AB.ab.
+
+This causes a dependency chain ``b.v -> ab.v -> aa.v``. Now we might be
+interested in building theory ``B``, so all we have to do is run ``dune build
+B``. Dune will automatically build the theory ``A`` since it is a dependency.
+
+.. _running-coq-top:
+
+Running a Coq Toplevel
+----------------------
+
+Dune supports running a Coq toplevel binary such as ``coqtop``, which is
+typically used by editors such as CoqIDE or Proof General to interact with Coq.
```

Comment:
This one requires some explanation since it is a little confusing otherwise. @rlepigre addeda command called `dune coq top` to dune earlier this year. The intention was to get PG to interface with dune. As you might know, PG does not use coqidetop but rather a separate mode on coqtop directly. This command allows calling coqtop on a file, and rebuilding what is necessary. AFAIK it is working well for users using emacs/PG.

This isn't to say that the command name isn't confusing however. I hope that in the future, we can also support coqidetop using this command, which would allow for some editor integration with dune. I don't think this command is intended for use by regular users, rather just a technical backend for PG for the time being.

The command grouping is a little strange, but make sense from the point of view of dune. Subcommands define the scope of functionality of a program, so there was a choice between `dune coqtop` and `dune coq top`. The latter allows for more `coq` related commands to be included under this group. I have no plans for any coq commands in dune however, nor am I aware of any more.

Reply:
> @Zimmi48 It doesn't work unfortunately, though I am not sure what is wrong.

That is surprising, what have you tried?

My understanding is that `coqtop` and `coqidetop` are kind of compatible in terms of the command-line arguments they expect, so I would have expected `dune coq top --toplevel coqidetop some/file.v` to do the trick.

Reply:
Is the problem the one showcased in https://github.com/ocaml/dune/pull/5552 ?

We need to fix that one asap.

Reply:
I don't remember if it was related to that. I will need to test further.